### DISCUSSION TITLE
So, what is DERP and its "regions"?

### DISCUSSION THREAD

**Discussion by Winterhuman** [n0-computer/iroh#1275] [Q&A] [answered]
From what I understand, DERP is a STUN and TURN-style relaying protocol borrowed from Tailscale. What I don't understand is what DERP regions are and why, or if, they're required or recommended; what functionality does it provide?

**Comment by Winterhuman** [comment]
@b5 Friendly ping (not sure who to ping specifically about the DERP parts of the codebase)

**Comment by flub** [comment, marked as answer]
The current derp regions design is copied directly from tailscale, which is an entirely closed system.  We want to enable anyone to run their own derp servers and I don't think we've entirely figured out how derp regions should work in that respect.

Having said that, derp regions as they are now are:

- Each region can be thought off as one logical derp server (though there could be multiple meshed derp servers handling a region).
- Derp servers **do not mesh** between regions.
- Each iroh node has a **home** derp region (if derp is enabled).
  - It connects to this when it is accepting new connections.
- To contact an iroh node you need `(derp_region, peer_id)`.
  - Regadless of your own derp region, you connect to dial a peer you connect to the derp region of the peer.
  - This derp server will do all the connection establishment and holepunching for you.

However right now regions are only identified by a `u16` which is a bit poor for the goal of letting anyone run their own network I think.  We'll have to figure these out better.
